moveit_ros/moveit_node_interface/src/node_interface.cpp

```diff
@@ -0,0 +1,80 @@
+/*********************************************************************
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2022, PickNik Robotics
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of PickNik Robotics nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Paul Gesel */
+
+#include "node_interface/node_interface.h"
+
+namespace node_interface {
+
+    rclcpp::node_interfaces::NodeBaseInterface::SharedPtr NodeInterface::get_node_base_interface() const {
+        return node_handle_->get_node_base_interface();
+    }
+
+    rclcpp::node_interfaces::NodeTopicsInterface::SharedPtr NodeInterface::get_node_topics_interface() const {
+        return node_handle_->get_node_topics_interface();
+    }
+
+    rclcpp::node_interfaces::NodeParametersInterface::SharedPtr NodeInterface::get_node_parameters_interface() const {
+        return node_handle_->get_node_parameters_interface();
+    }
+
+    std::shared_ptr<rclcpp::Node> NodeInterface::get_rcl_node() const {
+        if (!has_lifecycle_node_) {
+            return node_handle_->get_rcl_node();
+        }
+        return nullptr;
```

Comment:
This should probably return a `std::option<std::shared_ptr<rclcpp::Node>>` instead if a nullptr here.

The reason is that if you expose this and the user depends on this then it is easy for them to make mistakes and cause a segfault.

My understanding is that you want to use IK solvers in a ros2_control plugin which has a LifecycleNode. If any IK solvers use this function and depend on it returning a Node, you can't use it in your ros2_control controller.

If you can't just remove this function you should use option because that will create much nicer errors when a user makes a mistake and assumes this works. If you return a nullptr and they dereference it is at best going to segfault.

Reply:
I agree, it should be an optional.